internal/venice-common/src/main/java/com/linkedin/venice/pubsub/adapter/kafka/TopicPartitionsOffsetsTracker.java

```diff
@@ -64,51 +55,81 @@ public TopicPartitionsOffsetsTracker(@Nonnull Duration offsetsUpdateInterval, Du
    * to be the largest offset amongst the records passed in.
    *
    * @param records consumed records
+   * @param kafkaConsumer from which to extract metrics
    */
   public void updateEndAndCurrentOffsets(
       ConsumerRecords<byte[], byte[]> records,
-      Map<MetricName, ? extends Metric> metrics) {
-    if (lastMetricsCollectedTime != null && LatencyUtils
-        .getElapsedTimeInMs(lastMetricsCollectedTime.toEpochMilli()) < offsetsUpdateInterval.toMillis()) {
+      Consumer<byte[], byte[]> kafkaConsumer) {
+    long lastMetricsCollectedTimeSnapshot = lastMetricsCollectedTime;
+    long currentTime = System.currentTimeMillis();
+    long elapsedTime = currentTime - lastMetricsCollectedTimeSnapshot;
+    if (elapsedTime < offsetsUpdateIntervalMs) {
       return; // Not yet
     }
-    lastMetricsCollectedTime = Instant.now();
+    synchronized (this) {
```

Comment:
Since this is called from `Consumer::poll` which is already called with a mutex/synchronized, do we need another `synchronized`?

Reply:
That is true. I guess it is defensive coding, in case this assumption changes, though that seems unlikely. I could remove the thread safety check then.